pkg/sql/executor.go

```diff
@@ -1604,45 +1604,50 @@ func (e *Executor) execStmtInOpenTxn(
 		if err != nil {
 			return
 		}
-
-		if txnState.State() == AutoRetry && txnState.isSerializableRestart() {
-			// If we just ran a statement synchronously, then the parallel queue
-			// would have been synchronized first, so this would be a no-op.
-			// However, if we just ran a statement asynchronously, then the
-			// queue could still contain statements executing. So if we're in a
-			// SerializableRestart state, we synchronize to drain the rest of
-			// the stmts and clear the parallel batch's error-set before
-			// restarting. If we did not drain the parallel stmts then the
-			// txn.Reset call below might cause them to write at the wrong
-			// epoch.
-			//
-			// TODO(nvanbenschoten): like in Session.Finish, we should try to
-			// actively cancel these parallel queries instead of just waiting
-			// for them to finish.
-			if parErr := session.synchronizeParallelStmts(session.Ctx()); parErr != nil {
-				// If synchronizing results in a non-retryable error, it takes priority.
-				if _, ok := parErr.(*roachpb.HandledRetryableTxnError); !ok {
-					err = parErr
-					return
+		// If we can still automatically retry the txn, and we detect that the
+		// transaction won't be allowed to commit because its timestamp has been
+		// pushed, we go ahead and retry now - if we'd execute further statements,
+		// we probably wouldn't be allowed to retry automatically any more.
+		// We want to propagate the error if either:
+		// 1.we're in AutoRetry, for the reason described above.
+		// 2.retryIntent is not set, because the client would not benefit from letting the transaction run more statements.
```

Comment:
line is over 100 chars

The comment is also badly structured; the two paragraphs are not well integrated. I suggest:
```
Generally, if the transaction is serializable and it has been pushed, we don't deal with it until commit time (when we return a retriable error). This is because we want the transaction to continue and lay down all the intents it wants, so that a future attempt is more likely to succeed. However, there are two situations when we do want to detect the push and restart early:
1. If we can still automatically retry the txn, we go ahead and retry now - if we'd execute further statements, we probably wouldn't be allowed to retry automatically any more.
2. If the client is not following the client-directed retries protocol, then technically there won't be any retries, so the client would not benefit from letting the transaction run more statements.
```